File: predictor_api/predictor_api/service/knockout_template_service.py
from uuid import UUID

from fastapi import HTTPException

from db_handler.db_handler.model.column import Column
from db_handler.db_handler.model.query_condition import QueryCondition
from db_handler.db_handler.model.query_condition_group import (
    QueryConditionGroup
)
from db_handler.db_handler.model.query_request import QueryRequest
from db_handler.db_handler.model.query_response import QueryResponse
from db_handler.db_handler.model.table import Table
from db_handler.db_handler.model.type.sql_operator import SqlOperator
from db_handler.db_handler.model.update_request import UpdateRequest
from db_handler.db_handler.service.database_query_service import (
    DatabaseQueryService
)
from db_handler.db_handler.util.store_constants import StoreConstants
from predictor_api.predictor_api.model.knockout_template import (
    KnockoutTemplate
)
from predictor_api.predictor_api.model.tournament_template import (
    TournamentTemplate
)
from predictor_api.predictor_api.util.predictor_constants import (
    PredictorConstants
)


class KnockoutTemplateService:
    """
    Service for performing knockout template-related actions.

    Attributes:
        __query_service (DatabaseQueryService): The database query service.
    """

    def __init__(self, database_query_service: DatabaseQueryService):
        """
        Initialise the KnockoutTemplateService.

        Args:
            database_query_service (DatabaseQueryService): The database query
                service.
        """
        self.__query_service = database_query_service

    def get_knockout_templates(self) -> list[KnockoutTemplate]:
        """
        Retrieve stored knockout templates.

        Returns:
            list[KnockoutTemplate]: The stored knockout templates.
        """
        query_request: QueryRequest = QueryRequest(
            table=Table.of(
                PredictorConstants.PREDICTOR_SCHEMA,
                KnockoutTemplate.TARGET_TABLE
            )
        )

        query_response: QueryResponse = (
            self.__query_service.retrieve_records(query_request)
        )

        return list(
            map(
                lambda record:
                KnockoutTemplate.model_validate(record),
                query_response.records
            )
        )

    def create_knockout_templates(
            self,
            knockout_templates: list[KnockoutTemplate]
    ) -> list[KnockoutTemplate]:
        """
        Create new knockout templates.

        Args:
            knockout_templates (list[KnockoutTemplate]): The new knockout
                templates to create.

        Returns:
            list[KnockoutTemplate]: The newly created knockout templates.
        """
        self.__query_service.update_records(
            UpdateRequest(
                operation=SqlOperator.INSERT,
                table=Table.of(
                    PredictorConstants.PREDICTOR_SCHEMA,
                    KnockoutTemplate.TARGET_TABLE
                ),
                records=list(
                    map(
                        lambda template:
                        template.model_dump(exclude_none=True),
                        knockout_templates
                    )
                )
            )
        )

        return knockout_templates

    def get_knockout_template_by_id(
            self,
            knockout_template_id: UUID) -> KnockoutTemplate:
        """
        Retrieve a single stored knockout template by its id.

        Args:
            knockout_template_id (UUID): The id of the knockout template to
                retrieve.

        Returns:
            KnockoutTemplate: The retrieved knockout template.
        """
        query_response: QueryResponse = (
            self.__query_service.retrieve_records(
                QueryRequest(
                    table=Table.of(
                        PredictorConstants.PREDICTOR_SCHEMA,
                        KnockoutTemplate.TARGET_TABLE
                    ),
                    conditionGroup=QueryConditionGroup.of(
                        QueryCondition.of(
                            Column.of(StoreConstants.ID),
                            knockout_template_id
                        )
                    )
                )
            )
        )

        if len(query_response.records) == 0:
            raise HTTPException(
                status_code=404,
                detail="No knockout templates found with a matching id."
            )

        return KnockoutTemplate.model_validate(
            query_response.records[0]
        )

    def delete_knockout_template_by_id(
            self,
            knockout_template_id: UUID) -> None:
        """
        Delete a single stored knockout template by its id.

        Args:
            knockout_template_id (UUID): The id of the knockout template to
                delete.
        """
        query_request: QueryRequest = QueryRequest(
            table=Table.of(
                PredictorConstants.PREDICTOR_SCHEMA,
                TournamentTemplate.TARGET_TABLE
            ),
            conditionGroup=QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of("knockoutTemplateId"),
                    knockout_template_id
                )
            )
        )

        query_response: QueryResponse = (
            self.__query_service.retrieve_records(query_request)
        )

        if query_response.recordCount > 0:
            raise HTTPException(
                status_code=409,
                detail="Cannot delete knockout template as it is part of an "
                       "existing tournament template."
            )

        self.__query_service.update_records(
            UpdateRequest(
                operation=SqlOperator.DELETE,
                table=Table.of(
                    PredictorConstants.PREDICTOR_SCHEMA,
                    KnockoutTemplate.TARGET_TABLE
                ),
                conditionGroup=QueryConditionGroup.of(
                    QueryCondition.of(
                        Column.of(StoreConstants.ID),
                        knockout_template_id
                    )
                )
            )
        )

File: src/main/java/com/kanemullett/service/KnockoutTemplateService.java
package com.kanemullett.service;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import com.kanemullett.model.Column;
import com.kanemullett.model.ImmutableQueryRequest;
import com.kanemullett.model.ImmutableUpdateRequest;
import com.kanemullett.model.KnockoutTemplate;
import com.kanemullett.model.QueryCondition;
import com.kanemullett.model.QueryConditionGroup;
import com.kanemullett.model.QueryRequest;
import com.kanemullett.model.QueryResponse;
import com.kanemullett.model.Table;
import com.kanemullett.model.TournamentTemplateRecord;
import com.kanemullett.model.UpdateRequest;
import com.kanemullett.model.type.SqlOperator;
import com.kanemullett.util.DatabaseConstants;
import com.kanemullett.util.PredictorConstants;

@Service
public class KnockoutTemplateService {

    private final DatabaseQueryService queryService;

    public KnockoutTemplateService(DatabaseQueryService queryService) {
        this.queryService = queryService;
    }

    public List<KnockoutTemplate> getKnockoutTemplates() {
        final QueryRequest<KnockoutTemplate> queryRequest = ImmutableQueryRequest.<KnockoutTemplate>builder()
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, KnockoutTemplate.TARGET_TABLE))
            .recordClass(KnockoutTemplate.class)
            .build();

        final QueryResponse<KnockoutTemplate> queryResponse = queryService.retrieveRecords(queryRequest);

        return queryResponse.getRecords();
    }

    public List<KnockoutTemplate> createKnockoutTemplates(List<KnockoutTemplate> knockoutTemplates) {
        final UpdateRequest<KnockoutTemplate> updateRequest = ImmutableUpdateRequest.<KnockoutTemplate>builder()
            .operation(SqlOperator.INSERT)
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, KnockoutTemplate.TARGET_TABLE))
            .records(knockoutTemplates)
            .build();

        queryService.updateRecords(updateRequest);

        return knockoutTemplates;
    }

    public KnockoutTemplate getKnockoutTemplateById(String knockoutTemplateId) {
        final QueryRequest<KnockoutTemplate> queryRequest = ImmutableQueryRequest.<KnockoutTemplate>builder()
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, KnockoutTemplate.TARGET_TABLE))
            .conditionGroup(QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of(DatabaseConstants.ID),
                    knockoutTemplateId
                )
            ))
            .recordClass(KnockoutTemplate.class)
            .build();

        final QueryResponse<KnockoutTemplate> queryResponse = queryService.retrieveRecords(queryRequest);

        if (queryResponse.getRecordCount() == 0) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "No knockout templates found with a matching id.");
        }

        return queryResponse.getRecords().get(0);
    }

    public void deleteKnockoutTemplateById(String knockoutTemplateId) {
        final QueryRequest<TournamentTemplateRecord> queryRequest = ImmutableQueryRequest.<TournamentTemplateRecord>builder()
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, TournamentTemplateRecord.TARGET_TABLE))
            .conditionGroup(QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of(TournamentTemplateRecord.KNOCKOUT_TEMPLATE_ID_COLUMN),
                    knockoutTemplateId
                )
            ))
            .recordClass(TournamentTemplateRecord.class)
            .build();

        final QueryResponse<TournamentTemplateRecord> queryResponse = queryService.retrieveRecords(queryRequest);

        if (queryResponse.getRecordCount() > 0) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "Cannot delete knockout template as it is part of an existing tournament template.");
        }

        final UpdateRequest<KnockoutTemplate> updateRequest = ImmutableUpdateRequest.<KnockoutTemplate>builder()
            .operation(SqlOperator.DELETE)
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, KnockoutTemplate.TARGET_TABLE))
            .conditionGroup(QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of(DatabaseConstants.ID),
                    knockoutTemplateId
                )
            ))
            .build();

        queryService.updateRecords(updateRequest);
    }
}
